README.md

# Leaflet Virtual Grid

You can use `new VirtualGrid` to generate simple, cacheable, grids of `L.LatLngBounds` objects you can use to query APIs. This lets you query APIs for smaller units and space and never make a call data in the same area twice.

# Usage with Events

```js
var vg = new VirtualGrid();

// listen for when new cells come into the view for the first time
vg.on("cellcreate", function(e){
  console.log(e.type, e);
});

// listen for when cells reenter the view
vg.on("cellenter", function(e){
  console.log(e.type, e);
});

// listen for when cells leave the view
vg.on("cellleave", function(e){
  console.log(e.type, e);
});

// add the grid to the map
vg.addTo(map);
```

# Usage as a Class

```js
var MyGrid = new VirtualGrid.extend({
  createCell: function(bounds, coords){
    console.log('create cell', bounds, coords);
  },

  cellEnter: function(bounds, coords){
    console.log('cell enter', bounds, coords);
  },

  cellLeave: function(bounds, coords){
    console.log('cell leave', bounds, coords);
  }
})

var thingWithGrid = new MyGrid().addTo(map);
```

# Options

```js
var vg = new VirtualGrid({
  cellSize: 512,
  updateInterval: 150
});
```

##### `updateInterval`

How often to update the grid. Defaults to `150`

##### `cellSize`

How big each cell is in pixels. Defaults to `512`

# Example

Here is what the grid looks like under the hood...

![Example](https://raw.github.com/patrickarlt/leaflet-virtual-grid/master/example.jpg)

Each rectangle would represent a call to an API or query to a data source. You would only make one request per cell so you not make repeat calls to areas like requesting all the data in a map view when a user performs a small pan.

# Credit

Most of this code is based on `L.Grid` from https://github.com/Leaflet/Leaflet/commit/670dbaac045c7670ff26198136e440be9c2bb3e5.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

### [Unreleased]

## [1.0.6] 2018-03-23

* cleaned up npm tarball and started exposing `pkg.module` for downstream ES dependents.

## [1.0.5]

* ES6 Leaflet imports (courtesy of [@finneganh](https://github.com/finneganh))

## [1.0.4]

* i wish i could explain it.

## [1.0.3]

* Work around Leaflet/Leaflet#4023 in Leaflet 1.0.0-beta.2.

## [1.0.2]

* Fix NPM release

### Fixed

* Fix NPM release

## [1.0.1]

### Changed

* grid no longer attempts to honor min/maxZoom specified by layer manager (more info [here](https://github.com/Esri/esri-leaflet/pull/660#issuecomment-151680624))

## [1.0.0]

### Added

* Support for Leaflet 1.0.0-beta

## [0.1.1]

### Fixed

* Added Leaflet to package dependencies.

## 0.1.0

First release.

[Unreleased]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v1.0.6...HEAD
[1.0.6]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v1.0.5...v1.0.6
[1.0.5]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v0.1.1...v1.0.0
[0.1.1]: https://github.com/patrickarlt/leaflet-virtual-grid/compare/v0.1.0...v0.1.1


test/virtual-grid.js
import VirtualGrid from '../src/virtual-grid.js';
import test from 'tape-catch';
import sinon from 'sinon';
import { map, point } from 'leaflet';

function createMap () {
  // create container
  var container = document.createElement('div');

  // give container a width/height
  container.setAttribute('style', 'width:500px; height: 500px;');

  // add contianer to body
  document.body.appendChild(container);

  return map(container);
}

function createMockGrid () {
  var MockGrid = VirtualGrid.extend({
    createCell: sinon.spy(),
    cellEnter: sinon.spy(),
    cellLeave: sinon.spy()
  });

  return new MockGrid();
}

test('should exist', function (t) {
  t.plan(1);
  t.ok(VirtualGrid);
});

test('should create cells based on the view of the map', function (t) {
  t.plan(1);

  var map = createMap().setView([0, 0], 1);
  var grid = createMockGrid();

  grid.on('cellsupdated', function () {
    t.ok(grid.createCell.getCall(0).args[1].equals(point([0, 0])));
    map.remove();
  });

  grid.addTo(map);
});

test('should create cells when the map zooms in', function (t) {
  t.plan(5);

  var map = createMap().setView([0, 0], 1);
  var grid = createMockGrid();

  grid.addTo(map);

  grid.on('cellsupdated', function () {
    t.ok(grid.cellLeave.getCall(0).args[1].equals(point([0, 0, 1])));

    t.ok(grid.createCell.getCall(1).args[1].equals(point([0, 0, 2])));
    t.ok(grid.createCell.getCall(2).args[1].equals(point([1, 0, 2])));
    t.ok(grid.createCell.getCall(3).args[1].equals(point([0, 1, 2])));
    t.ok(grid.createCell.getCall(4).args[1].equals(point([1, 1, 2])));

    map.remove();
  });

  map.zoomIn();
});

test('should create cells when the map is panned', function (t) {
  t.plan(2);

  var map = createMap().setView([0, 0], 4);
  var grid = createMockGrid();

  grid.addTo(map);

  grid.on('cellsupdated', function () {
    t.ok(grid.createCell.getCall(4).args[1].equals(point([5, 4, 4])));
    t.ok(grid.createCell.getCall(5).args[1].equals(point([4, 5, 4])));
    map.remove();
  });

  map.panBy([512, 512], {
    animate: false,
    duration: 0
  });
});


scripts/release.sh
#!/bin/bash

# config
VERSION=$(node --eval "console.log(require('./package.json').version);")
NAME=$(node --eval "console.log(require('./package.json').name);")

# build and test
npm test || exit 1

# checkout temp branch for release
git checkout -b gh-release

# run prepublish to build files
npm run prepublish

# force add files
git add dist -f

# commit changes with a versioned commit message
git commit -m "build $VERSION"

# push commit so it exists on GitHub when we run gh-release
git push https://github.com/patrickarlt/leaflet-virtual-grid gh-release

# create a ZIP archive of the dist files
zip -r $NAME-v$VERSION.zip dist

# run gh-release to create the tag and push release to github
gh-release --assets $NAME-v$VERSION.zip

# checkout master and delete release branch locally and on GitHub
git checkout master
git branch -D gh-release
git push https://github.com/patrickarlt/leaflet-virtual-grid :gh-release

# publish release on NPM
npm publish

src/virtual-grid.js
import {
  bounds,
  latLngBounds,
  point,
  Layer,
  setOptions,
  Util
} from 'leaflet';

var VirtualGrid = Layer.extend({

  options: {
    cellSize: 512,
    updateInterval: 150
  },

  initialize: function (options) {
    options = setOptions(this, options);
    this._zooming = false;
  },

  onAdd: function (map) {
    this._map = map;
    this._update = Util.throttle(this._update, this.options.updateInterval, this);
    this._reset();
    this._update();
  },

  onRemove: function () {
    this._map.removeEventListener(this.getEvents(), this);
    this._removeCells();
  },

  getEvents: function () {
    var events = {
      moveend: this._update,
      zoomstart: this._zoomstart,
      zoomend: this._reset
    };

    return events;
  },

  addTo: function (map) {
    map.addLayer(this);
    return this;
  },

  removeFrom: function (map) {
    map.removeLayer(this);
    return this;
  },

  _zoomstart: function () {
    this._zooming = true;
  },

  _reset: function () {
    this._removeCells();

    this._cells = {};
    this._activeCells = {};
    this._cellsToLoad = 0;
    this._cellsTotal = 0;
    this._cellNumBounds = this._getCellNumBounds();

    this._resetWrap();
    this._zooming = false;
  },

  _resetWrap: function () {
    var map = this._map;
    var crs = map.options.crs;

    if (crs.infinite) { return; }

    var cellSize = this._getCellSize();

    if (crs.wrapLng) {
      this._wrapLng = [
        Math.floor(map.project([0, crs.wrapLng[0]]).x / cellSize),
        Math.ceil(map.project([0, crs.wrapLng[1]]).x / cellSize)
      ];
    }

    if (crs.wrapLat) {
      this._wrapLat = [
        Math.floor(map.project([crs.wrapLat[0], 0]).y / cellSize),
        Math.ceil(map.project([crs.wrapLat[1], 0]).y / cellSize)
      ];
    }
  },

  _getCellSize: function () {
    return this.options.cellSize;
  },

  _update: function () {
    if (!this._map) {
      return;
    }

    var mapBounds = this._map.getPixelBounds();
    var cellSize = this._getCellSize();

    // cell coordinates range for the current view
    var cellBounds = bounds(
      mapBounds.min.divideBy(cellSize).floor(),
      mapBounds.max.divideBy(cellSize).floor());

    this._removeOtherCells(cellBounds);
    this._addCells(cellBounds);

    this.fire('cellsupdated');
  },

  _addCells: function (cellBounds) {
    var queue = [];
    var center = cellBounds.getCenter();
    var zoom = this._map.getZoom();

    var j, i, coords;
    // create a queue of coordinates to load cells from
    for (j = cellBounds.min.y; j <= cellBounds.max.y; j++) {
      for (i = cellBounds.min.x; i <= cellBounds.max.x; i++) {
        coords = point(i, j);
        coords.z = zoom;

        if (this._isValidCell(coords)) {
          queue.push(coords);
        }
      }
    }

    var cellsToLoad = queue.length;

    if (cellsToLoad === 0) { return; }

    this._cellsToLoad += cellsToLoad;
    this._cellsTotal += cellsToLoad;

    // sort cell queue to load cells in order of their distance to center
    queue.sort(function (a, b) {
      return a.distanceTo(center) - b.distanceTo(center);
    });

    for (i = 0; i < cellsToLoad; i++) {
      this._addCell(queue[i]);
    }
  },

  _isValidCell: function (coords) {
    var crs = this._map.options.crs;

    if (!crs.infinite) {
      // don't load cell if it's out of bounds and not wrapped
      var cellNumBounds = this._cellNumBounds;
      if (
        (!crs.wrapLng && (coords.x < cellNumBounds.min.x || coords.x > cellNumBounds.max.x)) ||
        (!crs.wrapLat && (coords.y < cellNumBounds.min.y || coords.y > cellNumBounds.max.y))
      ) {
        return false;
      }
    }

    if (!this.options.bounds) {
      return true;
    }

    // don't load cell if it doesn't intersect the bounds in options
    var cellBounds = this._cellCoordsToBounds(coords);
    return latLngBounds(this.options.bounds).intersects(cellBounds);
  },

  // converts cell coordinates to its geographical bounds
  _cellCoordsToBounds: function (coords) {
    var map = this._map;
    var cellSize = this.options.cellSize;
    var nwPoint = coords.multiplyBy(cellSize);
    var sePoint = nwPoint.add([cellSize, cellSize]);
    var nw = map.wrapLatLng(map.unproject(nwPoint, coords.z));
    var se = map.wrapLatLng(map.unproject(sePoint, coords.z));

    return latLngBounds(nw, se);
  },

  // converts cell coordinates to key for the cell cache
  _cellCoordsToKey: function (coords) {
    return coords.x + ':' + coords.y;
  },

  // converts cell cache key to coordiantes
  _keyToCellCoords: function (key) {
    var kArr = key.split(':');
    var x = parseInt(kArr[0], 10);
    var y = parseInt(kArr[1], 10);

    return point(x, y);
  },

  // remove any present cells that are off the specified bounds
  _removeOtherCells: function (bounds) {
    for (var key in this._cells) {
      if (!bounds.contains(this._keyToCellCoords(key))) {
        this._removeCell(key);
      }
    }
  },

  _removeCell: function (key) {
    var cell = this._activeCells[key];

    if (cell) {
      delete this._activeCells[key];

      if (this.cellLeave) {
        this.cellLeave(cell.bounds, cell.coords);
      }

      this.fire('cellleave', {
        bounds: cell.bounds,
        coords: cell.coords
      });
    }
  },

  _removeCells: function () {
    for (var key in this._cells) {
      var cellBounds = this._cells[key].bounds;
      var coords = this._cells[key].coords;

      if (this.cellLeave) {
        this.cellLeave(cellBounds, coords);
      }

      this.fire('cellleave', {
        bounds: cellBounds,
        coords: coords
      });
    }
  },

  _addCell: function (coords) {
    // wrap cell coords if necessary (depending on CRS)
    this._wrapCoords(coords);

    // generate the cell key
    var key = this._cellCoordsToKey(coords);

    // get the cell from the cache
    var cell = this._cells[key];
    // if this cell should be shown as isnt active yet (enter)

    if (cell && !this._activeCells[key]) {
      if (this.cellEnter) {
        this.cellEnter(cell.bounds, coords);
      }

      this.fire('cellenter', {
        bounds: cell.bounds,
        coords: coords
      });

      this._activeCells[key] = cell;
    }

    // if we dont have this cell in the cache yet (create)
    if (!cell) {
      cell = {
        coords: coords,
        bounds: this._cellCoordsToBounds(coords)
      };

      this._cells[key] = cell;
      this._activeCells[key] = cell;

      if (this.createCell) {
        this.createCell(cell.bounds, coords);
      }

      this.fire('cellcreate', {
        bounds: cell.bounds,
        coords: coords
      });
    }
  },

  _wrapCoords: function (coords) {
    coords.x = this._wrapLng ? Util.wrapNum(coords.x, this._wrapLng) : coords.x;
    coords.y = this._wrapLat ? Util.wrapNum(coords.y, this._wrapLat) : coords.y;
  },

  // get the global cell coordinates range for the current zoom
  _getCellNumBounds: function () {
    var worldBounds = this._map.getPixelWorldBounds();
    var size = this._getCellSize();

    return worldBounds ? bounds(
      worldBounds.min.divideBy(size).floor(),
      worldBounds.max.divideBy(size).ceil().subtract([1, 1])) : null;
  }
});

export default VirtualGrid;
